session/advisory_locks.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package session
+
+import (
+	"context"
+
+	"github.com/pingcap/tidb/parser/terror"
+)
+
+// Advisory Locks are the locks in GET_LOCK() and RELEASE_LOCK().
+// We implement them in TiDB by using an INSERT into mysql.advisory_locks
+// inside of a pessimistic transaction that is never committed.
+//
+// Each advisory lock requires its own session, since the pessimistic locks
+// can be rolled back in any order (transactions can't release random locks
+// like this even if savepoints was supported).
+//
+// We use referenceCount to track the number of references to the lock in the session.
+// A little known feature of advisory locks is that you can call GET_LOCK
+// multiple times on the same lock, and it will only be released when
+// the reference count reaches zero.
+
+type advisoryLock struct {
+	session        *session
+	referenceCount int
+}
+
+// IncrReferences increments the reference count for the advisory lock.
+func (a *advisoryLock) IncrReferences() {
+	a.referenceCount++
+}
+
+// DecrReferences decrements the reference count for the advisory lock.
+func (a *advisoryLock) DecrReferences() {
+	a.referenceCount--
+}
+
+// References returns the current reference count for the advisory lock.
+func (a *advisoryLock) ReferenceCount() int {
+	return a.referenceCount
+}
+
+// Close releases the advisory lock, which includes
+// rolling back the transaction and closing the session.
+func (a *advisoryLock) Close() {
+	_, err := a.session.ExecuteInternal(context.Background(), "ROLLBACK")
+	terror.Log(err)
+	a.session.Close()
+}
+
+// GetLock acquires a new advisory lock using a pessimistic transaction.
+// The timeout is implemented by using the pessimistic lock timeout.
+// We will never COMMIT the transaction, but the err indicates
+// if the lock was successfully acquired.
+func (a *advisoryLock) GetLock(lockName string, timeout int64) error {
+	_, err := a.session.ExecuteInternal(context.Background(), "SET innodb_lock_wait_timeout = %?", timeout)
+	if err != nil {
+		return err
+	}
+	_, err = a.session.ExecuteInternal(context.Background(), "BEGIN PESSIMISTIC")
+	if err != nil {
+		return err
+	}
+	_, err = a.session.ExecuteInternal(context.Background(), "INSERT INTO mysql.advisory_locks (lock_name) VALUES (%?)", lockName)
```

Comment:
There are 2 enhancements to be considered:

1. `GET_LOCK()` can be interrupted in MySQL like:
```
mysql> select get_lock('a', 50);
^C^C -- query aborted
ERROR 1317 (70100): Query execution was interrupted
``` 

2. The time of lock to be held is controlled by [max-txn-ttl](https://docs.pingcap.com/tidb/dev/tidb-configuration-file#max-txn-ttl), the default value is 1hr(it might be good enough for most cases of named lock?). If the default value is quite short, it will be a problem for named lock; if it is quite long, this will block the GC of TiDB.

Reply:
For (2) it sounds like a bug that innodb_lock_wait_timeout can exceed max-txn-ttl:

1073741824 seconds > 3600000 ms

Reply:
I agree (1) is required to have, although for Ctrl+C to work it requires https://docs.pingcap.com/tidb/dev/tidb-configuration-file#compatible-kill-query enabled. It looks like `kill` on a connection that is acquiring a lock is currently non-interruptable, so I will fix it.

For (2) I think we should fix this in https://github.com/pingcap/tidb/issues/33472 - we will also need to document that GET_LOCK() has a compatibility difference that lock waiting is only up to 1 hr, which I agree is fine for most use cases.

Reply:
I have a fix for the first issue: it just requires `context.TODO()` for the attached session. But I'll push it in an update with tests in a day or two.